src/imarker_robot_state.cpp

```diff
@@ -161,27 +157,47 @@ void IMarkerRobotState::setToCurrentState()
   planning_scene_monitor::LockedPlanningSceneRO scene(psm_);  // Lock planning scene
   (*imarker_state_) = scene->getCurrentState();
 
-  // TODO: move interactive markers and pose
+  // Set updated pose from robot state
+  for (std::size_t i = 0; i < arm_datas_.size(); ++i)
+    end_effectors_[i]->setPoseFromRobotState();
 
   // Show new state
   visual_tools_->publishRobotState(imarker_state_, color_);
 }
 
-bool IMarkerRobotState::setToRandomState()
+bool IMarkerRobotState::setToRandomState(double clearance)
 {
-  ROS_ERROR_STREAM_NAMED(name_, "Not reimplemented yet");
-  /*
   static const std::size_t MAX_ATTEMPTS = 1000;
-  for (std::size_t i = 0; i < MAX_ATTEMPTS; ++i)
+  for (std::size_t attempt = 0; attempt < MAX_ATTEMPTS; ++attempt)
   {
-    imarker_state_->setToRandomPositions(jmg_);
+    // Set each planning group to random
+    for (std::size_t i = 0; i < arm_datas_.size(); ++i)
+    {
+      imarker_state_->setToRandomPositions(arm_datas_[i].jmg_);
+    }
+
+    // Update transforms
     imarker_state_->update();
+    planning_scene_monitor::LockedPlanningSceneRO planning_scene(psm_);  // Read only lock
 
-    if (isStateValid())
+    // Collision check
+    // which planning group to collision check, "" is everything
+    static const bool verbose = false;
+    if (planning_scene->isStateValid(*imarker_state_, "", verbose))
     {
-      // ROS_DEBUG_STREAM_NAMED(name_, "Found valid random robot state after " << i << " attempts");
-
-      // Get pose from robot state
+      // Check clearance
+      if (clearance > 0)
+      {
+        // which planning group to collision check, "" is everything
+        if (planning_scene->distanceToCollision(*imarker_state_) > clearance)
+        {
+          continue; // clearance is not enough
```

Comment:
If distanceToCollision > clearance than shouldn't this comment read "clearance is enough"?